### ISSUE TITLE
Admin standards specification

### ISSUE THREAD

**Issue by happyDemon** [modulargaming/modulargaming#79]
@hinton and me got to talking about this, better to discuss this in an appropriate ticket

**Mobile support**
Oscar brought up the issue of modals possible being a bother on mobile devices, it wouldn't be the best user-experience on mobile devices.

This can easily be fixed by offer a mobile tweak of the admin through a specialised user interface framework.

There are lots of options, but I'd personally go for jqMobi (reason: http://www.youtube.com/watch?v=MwNdWZsRXgk).

if the admin were to be visited through a mobile device we'd switch on jqMobi to handle the page loads (standard pages would still be sent as HTML, while the modals would be parsed by Mustache.js on the client side).

**Client-side template support**
The javascript flavour of mustache would be used here. Oscar had some concerns about precompiling mustache templates into javascript files.

We could write a minion task that reads out templates from the folder admin/[module]/modals/, converts them to one line of code and stores them in an array (which would be located in assets/templates/admin/[module]/pages.js)

**Automation of handling policies**
in the admin's controller we could automatically check for policies in a before method

``` php
$route = 'Admin_'.$this->request->controller().'_'.$this->request->action();
if(!$user->can($route) {
 //error
}
```

**Handling ajax requests**
to make this easier I'd love to implement something called restful controllers, this is a concept I learned about thanks to the FuelPHP framework, but Laravel has a nice hybrid implementation we could borrow ideas from : http://laravel.com/docs/controllers#restful-controllers.

standard requests would use the normal Kohana naming conventions e.g. every route is prefixed with action_.

Ajax requests, however, would get different prefixes (get_ for retrieving data, post_ for submitting data, I don't think we should implement delete_ or put_) any data passed on to this->response->body() would be encoded into json format and the response's content type headers will be set to json before sending the output.

We'd also need to find a way to generate CSRF tokens

**Interface unity**
We should discuss the basic style of the admin pages and how the game owners & moderators would interact with it.

**!!!IMPORTANT**
This is fully hypothical and should be still be discussed in detail.
After discussing we can seperate these points of interest into seperate tickets.
Any ideas welcome

**Comment by happyDemon** [comment]
I'm uploading a jQuery plugin tomorow that has all the CRUD functionality that the item admin page has, just a wrapper around the javascript code I had already uploaded.
It's portable and should make it easy to extend to other admin pages.
You would only need to specify the table that has the data, an empty form in a modal and a delete modal. callbacks would be optional ofcourse ^^

**Comment by Hinton** [comment]
I would prefer having the same website theme on both desktop and mobile devices, considering we are using Twitter bootstrap, we can easily support fluid design and adapt it to look nicely on smaller devices without much problem at all.
I never liked "mobile" themes as they tend to work bad and feel un-native while trying to look native.

The rest controller looks interesting and might be a good way to split up logic, will take a better look later today.

**Comment by curtisdelicata** [comment]
I agree that it is best to use the same website theme for all devices.

**Comment by Hinton** [comment]
Regarding templates, we should attempt to reuse the clientside and serverside, not sure the best approach for this.

One way would be to use a inner-layout with the modals content within.
Layout->layout inner->modal

**Comment by happyDemon** [comment]
that would be a viable option, at the moment I'm only using 2 modals for the item admin; a form modal and a delete confirmation modal (the delete modal wouldn't need to be recreated for every admin page)

**Comment by happyDemon** [comment]
It might come in handy include a few extra css style and js plugins in the admin:

http://favbulous.com/demo/twitter-bootstrap-custom-icons/
http://reactiveraven.github.com/jqBootstrapValidation/
http://www.eyecon.ro/bootstrap-datepicker/

**Comment by Hinton** [comment]
The famfamfam icons looks quite odd with bootstrap considering all the icons currently are black/white. The validation plugin looks nice, seems like it uses mostly html5 for validation.
Datepicker looks good and might be useful.

**Comment by happyDemon** [comment]
you're right there, I was just advising them for use in the admin's main nav bar

**Comment by happyDemon** [comment]
I'm also thinking about standardising the way the admin handles ajax requests.

This would mean we'd have to normalise what data the controller sends over the requests.
For example, all a successful request would be required to send send is a parameter called status with a value called success, any other data would be used to handle a callback.
While a request with errors would need 2 parameters; status: error, errors. :{[field: 'name', msg: ['error', 'error', 'error']]}
A callback would assign the errors to the appointed fields.

I would write a jQuery plugin to handle the client-side portion that just overloads the ajax loader.

**Comment by happyDemon** [comment]
https://gist.github.com/d20cd3eeb23b70502715

Here's a gist that describe how you'd generally set up an admin page (client side)

**Comment by curtisdelicata** [comment]
The gist looks good.

**Comment by happyDemon** [comment]
http://fortawesome.github.com/Font-Awesome/ we could maybe use some of those icons for the admin, offers us a broader selection for the nav an maybe even in the admin itself.

**Comment by happyDemon** [comment]
@hinton from having a look at the user admin page I'm guessing you would want to put in a complete ajax fallback? I thought we were just going to render the content of the forms into pages rather than modals for mobile support.

**Comment by happyDemon** [comment]
I'm thinking we could make use of the jQuery dataTables plugin, found an article that shows it's easy to integrate with bootstrap's css (http://datatables.net/blog/Twitter_Bootstrap).

This could come in handy for 90% of the admin pages, next to that most of the hard work could be handled by this module: https://github.com/morgan/kohana-datatables.

Next to that I found neat little image uploader that we could integrate in the pet & item admin: http://jasny.github.com/bootstrap/javascript.html#fileupload

**Comment by curtisdelicata** [comment]
dataTables and kohana-datatables looks interesting.

I was going to ask about if we were going to have image upload.

**Comment by happyDemon** [comment]
I think we should take the time for this in the next release (1.1.0), otherwise I don't see most of this being included or being rushed.

**Comment by curtisdelicata** [comment]
Mobile support:

We can support both because bootstrap should scale to the correct screen size? Do the Javascript prompts also scale?

Client-side template support:

This has been completed?

Automation of handling policies:

This is a good idea. What do you think @hinton 

Handling ajax requests:

This has been completed?

Interface unity:

Every module will have the same style of admin? (Using data tables and JS)

**Comment by Hinton** [comment]
Some notes for me to remember.

## CSRF

Easiest solution would be to include in template.

``` javascript
var crsf_token = {{crsf_token}};
```

It can then be included in the post by javascript, preferable using a custom .post method to prevent duplicated code.

## Actions

We should create or copy a Controller_REST that prepends the HTML verb, `action_index` would become `get_index`. As @happyDemon mentioned Laravel has a nice example, http://laravel.com/docs/controllers#restful-controllers.

## Policy

The optimal solution would be to handle Policy automatically, this could be performed by checking policy in before method in `Abstract_Controller_Admin`. Some things we have to be on the lookout for is, that we probably want to call the Policy again? with the correct data, for example in user admin, we do not want moderators to edit admins details.
